presto-main/src/main/java/com/facebook/presto/tracing/QueryStateTracingListener.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tracing;
+
+import com.facebook.presto.execution.QueryState;
+import com.facebook.presto.execution.StateMachine;
+import com.facebook.presto.spi.tracing.Tracer;
+
+import static java.util.Objects.requireNonNull;
+
+public class QueryStateTracingListener
+        implements StateMachine.StateChangeListener<QueryState>
+{
+    private QueryState previousState;
+    private final Tracer tracer;
+
+    public QueryStateTracingListener(Tracer tracer)
+    {
+        this.tracer = requireNonNull(tracer, "tracer is null");
+    }
+
+    @Override
+    public void stateChanged(QueryState newState)
+    {
+        synchronized (tracer) {
```

Comment:
Do we need to protect `previousState` as well? If yes, then we can make this method a synchronized completely